File: xiaoming/1011/FirstBadVersion.java

public class FirstBadVersion {
	
	public int firstBadVersionF(int n) {
        if (n < 1) return 0;
        
        int start = 1;
        int end = n;
        if (!isBadVersion(n)) return 0;

        while (start <= end) {
            int mid = (start + end ) / 2;
            if (isBadVersion(mid)) end = mid -1;
            else start = mid + 1;
        }
        return start;
    }

}

File: Sen/First_Bad_Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n): # iterative 
        """
        :type n: int
        :rtype: int
        """
        low, high = 1, n
        while low < high:
            mid = low + (high - low) / 2 # (high + low) may stack overflow
            if isBadVersion(mid):
                high = mid
            else:
                low = mid + 1
        return low # return high as the same as return low



    # def firstBadVersion(self, n): # recursive
    #     return self.binarySearch(1, n)
        
    # def binarySearch(self, low, high):
    #     if low >= high:
    #         return low
    #     mid = (low + high) / 2
    #     if isBadVersion(mid):
    #         return self.binarySearch(low, mid)
    #     else:
    #         return self.binarySearch(mid + 1, high)

